ipaserver/install/krbinstance.py

```diff
@@ -354,77 +435,111 @@ def _wait_for_replica_kdc_entry(self):
+            self.setup_full_pkinit()
         else:
-            subject = str(DN(('cn', self.fqdn), self.subject_base))
-            krbtgt = "krbtgt/" + self.realm + "@" + self.realm
-            certpath = (paths.KDC_CERT, paths.KDC_KEY)
-
-            try:
-                prev_helper = None
-                if self.master_fqdn is None:
-                    ca_args = [
-                        paths.CERTMONGER_DOGTAG_SUBMIT,
-                        '--ee-url', 'https://%s:8443/ca/ee/ca' % self.fqdn,
-                        '--certfile', paths.RA_AGENT_PEM,
-                        '--keyfile', paths.RA_AGENT_KEY,
-                        '--cafile', paths.IPA_CA_CRT,
-                        '--agent-submit'
-                    ]
-                    helper = " ".join(ca_args)
-                    prev_helper = certmonger.modify_ca_helper('IPA', helper)
-                else:
-                    self._wait_for_replica_kdc_entry()
-
-                certmonger.request_and_wait_for_cert(
-                    certpath,
-                    subject,
-                    krbtgt,
-                    dns=self.fqdn,
-                    storage='FILE',
-                    profile='KDCs_PKINIT_Certs')
-            except dbus.DBusException as e:
-                # if the certificate is already tracked, ignore the error
-                name = e.get_dbus_name()
-                if name != 'org.fedorahosted.certmonger.duplicate':
-                    root_logger.error("Failed to initiate the request: %s", e)
-                return
-            finally:
-                if prev_helper is not None:
-                    certmonger.modify_ca_helper('IPA', prev_helper)
-
-        # Finally copy the cacert in the krb directory so we don't
-        # have any selinux issues with the file context
-        shutil.copyfile(paths.IPA_CA_CRT, paths.CACERT_PEM)
+            self.setup_local_pkinit()
 
         try:
             self.restart()
         except Exception:
             root_logger.critical("krb5kdc service failed to restart")
             raise
 
+        record_pkinit_state_ldap(self.api)
+
     def test_anonymous_pkinit(self):
         with ipautil.private_ccache() as anon_ccache:
             try:
-                ipautil.run([paths.KINIT, '-n', '-c', anon_ccache])
-            except ipautil.CalledProcessError:
-                raise RuntimeError("Failed to configure anonymous PKINIT")
+                kinit_armor(anon_ccache, pkinit_anchor=paths.CACERT_PEM)
+            except ipautil.CalledProcessError as e:
+                raise RuntimeError(
+                    "Failed to configure anonymous PKINIT: {}".format(e))
 
     def enable_ssl(self):
         if self.config_pkinit:
             self.steps = []
             self.step("installing X509 Certificate for PKINIT",
                       self.setup_pkinit)
-            self.step("testing anonymous PKINIT", self.test_anonymous_pkinit)
-
             self.start_creation()
+        else:
```

Comment:
That's a question of UX we may discuss further. by default we inform the user that PKINIT certificate is being requested and installed. When he passes in `--no-pkinit` we pretend to not configure it and only use the local self-sign path. I thought that in this  case we shouldn't inform the user about this, otherwise we will have plenty complaints about an installer setting up undesired feature even when instructed not to.